server/src/main/java/org/elasticsearch/action/search/FetchSearchPhase.java

```diff
@@ -116,7 +116,8 @@ private void innerRun() throws Exception {
         // still use DFS_QUERY_THEN_FETCH, which does not perform the "query and fetch" optimization during the query phase.
         final boolean queryAndFetchOptimization = searchPhaseShardResults.length() == 1
             && context.getRequest().hasKnnSearch() == false
-            && reducedQueryPhase.queryPhaseRankCoordinatorContext() == null;
+            && reducedQueryPhase.queryPhaseRankCoordinatorContext() == null
+            && (context.getRequest().source() == null || context.getRequest().source().rankBuilder() == null);
```

Comment:
Yeap, this is a more generic bug (not active by any codepaths for now though). Will address in a different PR (prob with the semantic reranker rework, as this will be needed there)